### DISCUSSION TITLE
Gets the user selected text ?

### DISCUSSION THREAD

**Discussion by zsio** [tauri-apps/tauri#5624] [Q&A] [answered]
How to get user selected text in other app ？

For example, select text in the browser 。

**Comment by zsio** [comment]
![image](https://user-images.githubusercontent.com/16083606/201641216-ad502a32-5ad6-4a78-b4bb-7167f0c0fcd9.png)

**Comment by FabianLars** [comment, marked as answer]
> How to get user selected text in **other** app?

I don't think that this is possible at all. At least not if you don't target a specific app where you have access to stuff like this.

Apps and gui frameworks generally draw this themselves without necessarily telling the OS that text is selected or providing an API that other apps can call to get selected text.

The only thing that could be possible is to take a screenshot of the display and analyze that - this will very likely be really difficult, with many false detections.

In other words: Not possible.

**Reply by zsio** [reply]
https://www.npmjs.com/package/electron-selected-text

It can do that

**Reply by FabianLars** [reply]
Huh, interesting! Didn't expect that.

Anyway, you can use robotjs in a nodejs [sidecar](https://tauri.app/v1/guides/building/sidecar) then!

**Reply by zsio** [reply]
But I want to use it in tauri

**Reply by FabianLars** [reply]
Yeah, but unless there is a similar Rust crate (no idea if there is one) you have to bundle an additional binary that can do it (be it nodejs, or python or whatever else you find).

**Reply by zsio** [reply]
I see how he did it.

1、Save the contents of the current clipboard
2、Clear the contents of the clipboard
3、Let the system trigger the copy shortcut
4、Read the copied content
5, restore the initial contents of the clipboard

**Reply by FabianLars** [reply]
that's so smart and simple, didn't even think about using the keyboard shortcut 😅 

For input simulation there are crates like enigo, and for reading the clipboard crates like arboard or tauri's built-in clipboard manager (which will be backed-up by arboard in the future)

**Reply by zsio** [reply]
https://docs.rs/enigo/0.0.14/enigo/

This might work！ :)  

Thank you for your reply.

I'm closing this discussion. thanks again.

**Reply by jackeydou** [reply]
Hi, I just have a crash in my app by using enigo to simulate command+c, have you ever met this problem? I'm looking for the reason all day.

**Reply by zsio** [reply]
I'm sorry, I was just looking for options

**Reply by geekyayush** [reply]
Have you guys figured it out? :) especially for windows?

**Comment by seidtgeist** [comment]
For macOS, I've worked around having to install an extension by using AppleScript:

``` ts
import { readText, writeText } from "@tauri-apps/api/clipboard";
import { register, isRegistered } from "@tauri-apps/api/globalShortcut";
import { Command } from "@tauri-apps/api/shell";

const script = `
tell application "System Events"
    keystroke "c" using {command down}
end tell
`;

const args = script
  .trim()
  .split("\n")
  .flatMap((line) => ["-e", line]);

const command = new Command("osascript", args);

async function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function getSelectedText() {
  const initialClipboardContent = await readText();

  try {
    await command.spawn();

    await new Promise((resolve) => {
      command.on("close", resolve);
    });

    let attempts = 10;
    let currentClipboardContent = initialClipboardContent;

    while (
      attempts > 0 &&
      currentClipboardContent === initialClipboardContent
    ) {
      await sleep(100);
      currentClipboardContent = await readText();
      attempts--;
    }

    return currentClipboardContent;
  } finally {
    if (initialClipboardContent) {
      await writeText(initialClipboardContent);
    }
  }
}

async function main() {
  const shortcut = "Command+E";
  if (!(await isRegistered(shortcut))) {
    await register(shortcut, async () => {
      const text = await getSelectedText();

      document.body.innerHTML = text || "—";
    });
  }
}

main();
```

**Reply by HuakunShen** [reply]
Hi, I wonder whether your app window is in focus when this code is run? 
If your app is in focus then how to copy text from another app?
Oh I see, you are using a global hotkey so no window is in focus.
